Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [23]:
NAME = "Richard Verloop, #2013143#"
COLLABORATORS = "Stijn Meijs, #2015295#"

---

Fill in the cell above to provide us with your name and student number, like

NAME = "Adam Smith, #student number#"

where you replace "#student number#" with your ... (very good!)

and if you are doing this with someone else (not more than one other person):

COLLABORATORS = "Karl Marx, #student number#"

# exam June 7th, 2018



In the first cell of the notebook, give us your name and student number in the way indicated above. If you do this notebook with --at max.-- one other student, give her/his name and student number as well under "COLLABORATORS".

Fill in the notebook (see below for code cells and text cells that you need to fill in).

If you look at the menus above (File, Edit, View etc.), there is one called "Cell". If you click on this, you can change the "Cell Type". Choose "Code" when you are typing python or R code. Choose "Markdown" when you are typing, well, markdown.

When you finish the notebook, make sure that you **save it with the output of your code included**. 

Then put it on github, e.g. by dragging it onto github (see instructions below). 

Finally, add a link to your README file with the name of this exam: "Exam June 7, 2018".



## price effects of mergers

We consider a merger in this notebook. 

First we calculate the equilibrium with three firms, denoted by 1, 2 and 3. Then firms 2 and 3 merge so that we are left with 2 firms; denoted by 1 and 2.

We calculate the effects of the merger on the equilibrium price.

We start by importing the usual libraries.

In [94]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from scipy import optimize
%matplotlib inline

The following code helps to calculate the equilibrium for the case with 3 firms. Note that you have seen similar code during the tutorials to calculate the equilibrium with 2 firms.

We assume that before the merger each firm has constant marginal costs equal to 0.3. We assume a simple linear (inverse) demand curve of the form $p=1-Q$ where $p$ denotes price and $Q$ total output on the market. Total output equals the sum of each firm's output: $Q= q_1 + q_2+q_3$.

In [102]:
c0 = 0.3
vector_c = [c0]*3

def p(Q):
    return 1 - Q

def costs(q,c):
    return c*q

def profits(q,Q_other,c):
    return p(q+Q_other)*q-costs(q,c)

def reaction(Q_other,c):
    q1 =  optimize.fminbound(lambda x: -profits(x,Q_other,c),0,1,full_output=1)
    return q1[0]

def fixed_point_three_firms(vector_q,vector_c):
    return [vector_q[0]-reaction(vector_q[1]+vector_q[2],vector_c[0]),
            vector_q[1]-reaction(vector_q[0]+vector_q[2],vector_c[1]),
            vector_q[2]-reaction(vector_q[0]+vector_q[1],vector_c[2])]



#### a) [0.5 points] Show that total equilibrium output equals  0.525. Derive the equilibrium price and Herfindahl index; denote these by `P0` and `H0` resp.

In [103]:
initial_guess = [0,0,0]

sum(optimize.fsolve(lambda q: fixed_point_three_firms(q, vector_c), initial_guess))



0.525

In [104]:
P0 = p(0.525)
print(P0)

0.475


In [115]:

def Herfindahl_index(q_1,q_2,q_3):
    return ((q_1/Q)**2 + (q_2/Q)**2 + (q_3/Q)**2)

q_1 = 0.175
q_2 = 0.175
q_3 = 0.175
Q= 0.525


H0 = Herfindahl_index(q_1,q_2,q_3)

print(H0)

0.3333333333333333


#### b) [0.5 point] Define a function `fixed_point_two_firms` with the same structure as the function `fixed_point_three_firms` above, except that it derives the equilibrium output levels for a duopoly (two firms). Test this function by showing that each of the two firms produces 0.3333 in case both firms have zero costs.

In [116]:
def fixed_point_two_firms(vector_q, vector_c):
    return[vector_q[0]-reaction(vector_q[1],vector_c[0]),
        vector_q[1]-reaction(vector_q[0],vector_c[1])]
          
initial_guess = [0,0]

optimize.fsolve(lambda q: fixed_point_two_firms(q,[0.3, 0.6]), initial_guess)
                

array([0.33333333, 0.03333333])

The Dutch competition authority, ACM, is asked to evaluate the effects of a merger between firms 2 and 3. Firms 2 and 3 claim that by merging they can reduce their constant marginal costs. But it is not clear by how much they will reduce their costs.

The ACM assumes that the marginal cost level of the merged firm is uniformly distributed between 0 and the current marginal cost level `c0`. The merger will not affect the marginal cost level of firm 1 who does not merge. Firm 1's cost level remains `c0`.

The next cell generates a vector of cost levels for the merged firm, denoted `c_after_merger`. Evaluate "c_after_merger" in a separate cell if you do not understand the command that defines it.

Then it calculates the equilibrium output levels for (the non-merging) firm 1 and (the merged) firm 2.

Note that you need to specify `initial_guess`.


In [117]:
c_after_merger = np.random.uniform(0,c0,size = 100)

q1_after_merger = [optimize.fsolve(lambda q: fixed_point_two_firms(q,[c0,c]), initial_guess)[0] for c in c_after_merger]
q2_after_merger = [optimize.fsolve(lambda q: fixed_point_two_firms(q,[c0,c]), initial_guess)[1] for c in c_after_merger]




#### c) [0.5 points] Create a dataframe called `df_after_merger` with three columns: `c_merged_firm`, `output_non_merging_firm`, `output_merged_firm` containing resp. the cost level of the merged firm, the output level of firm 1 and the output level of firm 2.

In [151]:
initial_guess = [0.3, 0.15]

columns = ["c_merged_firm", "output_non_merging_firm", "output_merged_firm"]
df_after_merger = pd.DataFrame(columns=columns)

counter = 0
for i in range (99):
    c_merged_firm = c_after_merger[counter]
    df_after_merger.loc[counter] = [c_merged_firm, q1_after_merger, q2_after_merger]
    counter = counter + 1

df_after_merger.head()



c_merged_firm                            output_non_merging_firm  \
0       0.059283  [0.15309446988020525, 0.1722927069061322, 0.15...   
1       0.116878  [0.15309446988020525, 0.1722927069061322, 0.15...   
2       0.079564  [0.15309446988020525, 0.1722927069061322, 0.15...   
3       0.010012  [0.15309446988020525, 0.1722927069061322, 0.15...   
4       0.146320  [0.15309446988020525, 0.1722927069061322, 0.15...   

                                  output_merged_firm  
0  [0.39381106023958995, 0.35541458618773564, 0.3...  
1  [0.39381106023958995, 0.35541458618773564, 0.3...  
2  [0.39381106023958995, 0.35541458618773564, 0.3...  
3  [0.39381106023958995, 0.35541458618773564, 0.3...  
4  [0.39381106023958995, 0.35541458618773564, 0.3...

#### d) [0.5 points] Add three columns to the dataframe with resp. total equilibrium output on the market, `Q`, equilibrium price, `P` and the Herfindahl index, `H`.

In [161]:
df_after_merger['Q'] = df_after_merger['output_non_merging_firm'] + df_after_merger['output_merged_firm']
df_after_merger['P'] = 1 - df_after_merger['Q']
df_after_merger['H'] = (df_after_merger['output_non_merging_firm']/df_after_merger['Q'])**2 + (df_after_merger['output_merged_firm'/df_after_merger['Q'])**2

df_after_merger.head()

SyntaxError: invalid syntax (<ipython-input-161-1a6ac2d50d9f>, line 3)

#### e) [0.5 points] Make a histogram of the equilibrium price `P` after the merger. Also indicate in the histogram the equilibrium price before the merger `P0`. Label the horizontal axis with $P$.

[hint: you may want to use matplotlib's `hist`, `vlines` and `legend` to make this graph (e.g use google to find these functions); but feel free to use something else]

In [153]:
plt.hist(df_after_merger['P'], bins= 15, color='red')
plt.hist(P0, range=(0.43, 0.53))
plt.title('Equilibrium price')
plt.xlabel('P')
plt.ylabel('Frequency')
plt.show
           

KeyError: 'P'

#### f) [0.5 points] Explain why sometimes the equilibrium price after the merger exceeds the equilibrium price before the merger and sometimes it is lower than the pre-merger price. [hint: there is a trade off between two effects, depending on which effect dominates, equilibrium price increases or decreases after the merger]

It depends how many players are willing to contribute,
if few contribute, than the new merger's price will be higher than the efficient equilibrium price. Likewise, the more mergers contribute the lower the costs and therefor lower than the equilibrium price.
The problem being that not contributing will result in lower overall costs but stable income, which will then lead to higher mergers prices. This phenomenon is called free riding.

#### g) [0.5 points] Make a graph with the Herfindahl index on the horizontal axis and the equilibrium price on the vertical axis. This is straightforward for $(H,P)$ after the merger as both values are in the dataframe. Add in another color, the pre-merger combination `(H0,P0)` that we calculated above.

#### h) [0.5 points] Use the graph under g) to discuss the relation between the Herfindahl index and the equilibrium price. To illustrate, some people think that lower values of the Herfindahl index are associated with more competitive outcome. Would you agree with this?

The Herfindahl index exists of the squared sum of percentages. If there is a bigger concentration of total revenue from one supplier(lower competition) , the index will be higher. So we agree with the statement that lower values of the index are assi

## Github

After you have finished, we need to upload this notebook on github. Make sure that you upload the file on the github page of each collaborator.

Instructions on how to upload this on github can be found [on this page](http://janboone.github.io/programming-for-economists/github.html). This page has two screencasts: one shows how to drag the notebook onto your github page, the other shows how you can use the command line to upload your notebook.


Remember to update the README file in your repository to include a link to this notebook on github.


The links that you should post start with “github.com/” and are NOT of the form “http://localhost”. Make sure you test your links after uploading.